### ISSUE TITLE
Replicated control planes regression bug

### ISSUE THREAD

**Issue by pupseba** [istio/istio#27909]
Following [instructions ](https://istio.io/latest/docs/setup/install/multicluster/gateways/)to configure replicated planes leads to "503: service unavailable" when trying to communicate between pods cross-cluster. This happens with versions 1.7.x. Instructions work as expected with version 1.6.8. The issue has been reported by other users in places like https://github.com/istio/istio/issues/25077.

Curl test result:
```
kubectl exec --context=context1 $SLEEP_POD -n multi-test -c sleep -- curl -k -I httpbin.multi-test-bar.global:8000/headers
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0    91    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
HTTP/1.1 503 Service Unavailable
content-length: 91
content-type: text/plain
date: Wed, 07 Oct 2020 13:55:19 GMT
server: envoy
```

Istio-proxy container in sleep pod on origin cluster:
```
[2020-10-07T13:58:21.775Z] "HEAD /headers HTTP/1.1" 503 UF,URX "-" "-" 0 0 137 - "-" "curl/7.69.1" "5ccf05b7-d0e3-9e38-a581-8c0bdabc98b3" "httpbin.multi-t0" "10.14.10.99:31383" outbound|8000||httpbin.multi-test-bar.global - 240.0.0.2:8000 172.17.141.20:59704 - default
```

Ingress pods on destination cluster
```
[2020-10-07T13:58:21.900Z] "- - -" 0 NR "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 172.17.184.60:15443 172.31.4.248:62395 - -
[2020-10-07T13:58:21.814Z] "- - -" 0 NR "-" "-" 0 0 0 - "-" "-" "-" "-" "-" - - 172.17.133.59:15443 172.31.4.209:38326 - -
```

[X] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Following instructions works.

**Steps to reproduce the bug**
Follow instructions explained in this doc. https://istio.io/latest/docs/setup/install/multicluster/gateways/ with version 1.7.x.

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
istioctl version --remote
client version: 1.7.3
control plane version: 1.7.3
data plane version: 1.7.3 (28 proxies)

kubectl version --short
Client Version: v1.18.6
Server Version: v1.17.7

**How was Istio installed?**
Istio was installed using the Istio Operator in version 1.7.

**Environment where bug was observed (cloud vendor, OS, etc)**
We been able to reproduce the bug in different OS and environments that include. Physical, Virtual (vSphere) nodes with either CoreOS v31 or Fedora CoreOS v32.

**Comment by rytswd** [comment]
I was hit by this as well, but have managed to pull together a working configuration.

### Background

The multicluster documentation certainly needs correction, but this is rather an expected change following the [breaking change in EnvoyFilter configuration](https://istio.io/latest/news/releases/1.7.x/announcing-1.7/upgrade-notes/#envoyfilter-syntax-change).

This was discussed in https://github.com/istio/istio/issues/27204, and the solution is shared later in the [comment](https://github.com/istio/istio/issues/27204#issuecomment-704764102).
Given the behaviour change in Envoy, I believe this configuration is rather a long term fix than a workaround. This should be a part of EnvoyFilter handling when using `istioctl`, so in that sense a proper fix is for `istioctl` to generate the appropriate configuration from the start.

### (Interim) Solution

_NOTE_: As mentioned above, the configuration itself should be considered as a long term fix rather than an interim one.

You'd need to update `istio-multicluster-ingressgateway` EnvoyFilter CR in the destination cluster. You may want to check the CR deployed to the cluster first.
(If you are following along the official document, add `--context=$CTX_CLUSTER2` at the end of the command.)

```bash
$ kubectl get envoyfilter -n istio-system istio-multicluster-ingressgateway -o yaml

apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
# ... snip ...
  name: istio-multicluster-ingressgateway
  namespace: istio-system
# ... snip ...
spec:
  configPatches:
  - applyTo: NETWORK_FILTER
    match:
      context: GATEWAY
      listener:
        filterChain:
          filter:
            name: envoy.filters.network.sni_cluster
        portNumber: 15443
    patch:
      operation: INSERT_AFTER
      value:
        config:
          cluster_pattern: \.global$
          cluster_replacement: .svc.cluster.local
        name: envoy.filters.network.tcp_cluster_rewrite
  workloadSelector:
    labels:
      istio: ingressgateway
```

Given the above YAML, you can make the following change - the easiest would be to run `kubectl edit envoyfilter -n istio-system istio-multicluster-ingressgateway`.
(Again, if you are following along the official document, add `--context=$CTX_CLUSTER2` at the end of the command.)

_NOTE_: If you are using Istio Operator install, `kubectl edit` won't work as Istio Operator will override manual edits. You can instead create a separate EnvoyFilter CR in that case.

**BEFORE**

```yaml
spec:
  configPatches:
  - applyTo: NETWORK_FILTER
    # ... snip ...
    patch:
      operation: INSERT_AFTER
      value:
        # 👇 FROM HERE
        config:
        # 👆 TO HERE
          cluster_pattern: \.global$
          cluster_replacement: .svc.cluster.local
        name: envoy.filters.network.tcp_cluster_rewrite
```

**AFTER**

```yaml
spec:
  configPatches:
  - applyTo: NETWORK_FILTER
    # ... snip ...
    patch:
      operation: INSERT_AFTER
      value:
        # 👇 FROM HERE
        typed_config:
          "@type": "type.googleapis.com/istio.envoy.config.filter.network.tcp_cluster_rewrite.v2alpha1.TcpClusterRewrite"
        # 👆 TO HERE
          cluster_pattern: \.global$
          cluster_replacement: .svc.cluster.local
        name: envoy.filters.network.tcp_cluster_rewrite
```

### Upgrade Istio from v1.6 to v1.7

As detailed in the [v1.7 upgrade note](https://istio.io/latest/news/releases/1.7.x/announcing-1.7/upgrade-notes/#envoyfilter-syntax-change), if you are upgrading from v1.6 to v1.7, you may want to have both v1.6 compatible and v1.7 compatible EnvoyFilter CRs while upgrade is taking place. You can control the Envoy version for EnvoyFilter to be applied by adding `proxy` stanza like below:

```yaml
spec:
  configPatches:
  - applyTo: NETWORK_FILTER
    match:
      context: GATEWAY
      # 👇 FROM HERE
      proxy:
        proxyVersion: ^1\.7.*
      # 👆 TO HERE
    # ... snip ...
    patch:
      operation: INSERT_AFTER
      value:
        # 👉 This is for v1.7
        typed_config:
          "@type": "type.googleapis.com/istio.envoy.config.filter.network.tcp_cluster_rewrite.v2alpha1.TcpClusterRewrite"
        # 👈
          cluster_pattern: \.global$
          cluster_replacement: .svc.cluster.local
        name: envoy.filters.network.tcp_cluster_rewrite
```

You'd need to handle these configurations outside of `istioctl` setup, but hopefully this is just one more configuration to handle in addition to custom VS / GW / DR etc.

I haven't had a chance to test in Istio control plane canary upgrade scenario, so cannot speak about that unfortunately.

**Comment by pupseba** [comment]
Hi @rytswd,

Modifying the "config" for "typed_config" with the proper "@type" as shown in your entry worked perfectly in my case.

Documentation needs an upgrade yes, but shouldn't also istio-multicluster-ingressgateway resource be configured by default with "typed_config" instead of using "config" when multicluster is enabled? If that was the case, documentation should only mention this as part for troubleshooting and not as part of regular config.

**Comment by rytswd** [comment]
Hi @pupseba,

Yes, this certainly needs a fix for installation path - which actually seems to be already in place with https://github.com/istio/istio/pull/27264. I don't think this is being cherry-picked back to 1.7 branch, though. I have raised it to the release managers in Istio Slack (it should be fairly straightforward). I think v1.7.4 will be released soon, so this fix will only make it into v1.7.5 at the earliest.

Also, as to the documentation, the multicluster documentation seems to be getting a massive rewrite and restructuring in v1.8. You can check out how it's going to look at https://preliminary.istio.io/latest/docs/setup/install/multicluster.

**Comment by mandarjog** [comment]
@stevenctl can you follow up on fixing it in 1.7.x?

**Comment by stevenctl** [comment]
Does https://github.com/istio/istio/pull/28283 not fix this for 1.7?

**Comment by rytswd** [comment]
I think it should. I thought this was just pending the next release from 1.7 branch

**Comment by rytswd** [comment]
The fix has been deployed with 1.7.5, and I could confirm this works following the upgrade 👍
